Add global asset cache

IAmTrial opened the pull request:
Implements a global cache that loads all of the required files (as determined by calls to LoadFileInMem) into memory. This is intended to centralize the task of file loading and unloading to a singleton data structure. This changes file reads to batch all of the files at once, rather than spread it out to other files. These files are used in a read-only manner, as the buffer to store the files in memory are never modified.

In terms of tracking global usage and dynamic memory allocation, this is easier to manage, compared to the scattered logic across multiple source files. The cache loads files at the start of the program and unloads at the end of the program.

This is especially important for parallel computing device code, where the host is responsible for allocating and deallocating dynamic memory prior to calling device code. Device code cannot dynamically allocate memory. Multithreaded programs also benefit from not having to handle multiple threads reading the same read-only file.

The program may occupy more memory throughout its entire lifetime, but gains a speed advantage of not needing to read the same file in certain scenarios.

IAmTrial committed:
Implement global asset cache

This cache is intended to load and unload assets from a unified data
structure. This helps with moving all LoadFileInMem function (which is a
dynamic memory allocation function) into a well-defined sequence in the
code.

This change only introduces the global cache. It does not replace the
existing Diablo LoadFileInMem behavior.

IAmTrial committed:
Replace LoadFileInMem calls w/ global asset cache

AJenbo commented:
There is no need to load all this data, could we please limit it to the few relevant files:

```
levels/l1data/banner1.dun
levels/l1data/banner2.dun
levels/l1data/l1.min
levels/l1data/l1.sol
levels/l1data/l1.til
levels/l1data/rnd6.dun
levels/l1data/skngdo.dun
levels/l2data/blind2.dun
levels/l2data/blood2.dun
levels/l2data/bonestr1.dun
levels/l2data/l2.min
levels/l2data/l2.sol
levels/l2data/l2.til
levels/l3data/anvil.dun
levels/l3data/l3.min
levels/l3data/l3.sol
levels/l3data/l3.til
levels/l4data/diab1.dun
levels/l4data/diab2a.dun
levels/l4data/diab2b.dun
levels/l4data/diab3a.dun
levels/l4data/diab3b.dun
levels/l4data/diab4a.dun
levels/l4data/diab4b.dun
levels/l4data/l4.min
levels/l4data/l4.sol
levels/l4data/l4.til
levels/l4data/warlord2.dun
levels/l4data/warlord.dun
```

Note that this requires replacing the following for there alternatives:
```
blood1.dun -> blood2.dun
blind1.dun -> blind2.dun
bonestr2.dun -> bonestr1.dun
```

Besides that you can also load the `vile1.dun` though it's not being accessed by the current code as it only deals with single player.

Likewise Hellfire is not currently supported, but if you feel it's better to add the code for it then you can load:
```
nlevels/l5data/l5.min
nlevels/l5data/l5.sol
nlevels/l5data/l5.til
nlevels/l5data/l6.min
nlevels/l5data/l6.sol
nlevels/l5data/l6.til
```

But these should not be required for operation.

Also I would prefer if the application keeps working with just the shareware date and  the other files are optional:
```
levels/l1data/l1.min
levels/l1data/l1.sol
levels/l1data/l1.til
```

Making these changes should reduce the global cache to about 128kb which isn't to much more memory.

AJenbo commented:
Here are the files grouped in a way that may make more logical sens for the implementation:

Shareware/required
```
levels/l1data/l1.min
levels/l1data/l1.sol
levels/l1data/l1.til
```

Retail
```
levels/l1data/rnd6.dun
levels/l2data/l2.min
levels/l2data/l2.sol
levels/l2data/l2.til
levels/l3data/l3.min
levels/l3data/l3.sol
levels/l3data/l3.til
levels/l4data/diab1.dun
levels/l4data/diab2a.dun
levels/l4data/diab2b.dun
levels/l4data/diab3a.dun
levels/l4data/diab3b.dun
levels/l4data/diab4a.dun
levels/l4data/diab4b.dun
levels/l4data/l4.min
levels/l4data/l4.sol
levels/l4data/l4.til
```

Retail Single Player
```
levels/l1data/banner1.dun
levels/l1data/banner2.dun
levels/l1data/skngdo.dun
levels/l2data/blind2.dun
levels/l2data/blood2.dun
levels/l2data/bonestr1.dun
levels/l3data/anvil.dun
levels/l4data/warlord2.dun
levels/l4data/warlord.dun
```

Retail Multi Player
```
levels/l4data/vile1.dun
```

Hellfire
```
nlevels/l5data/l5.min
nlevels/l5data/l5.sol
nlevels/l5data/l5.til
nlevels/l5data/l6.min
nlevels/l5data/l6.sol
nlevels/l5data/l6.til
```

IAmTrial committed:
Restrict asset caching with cache options

This not only adds options for the asset cache loading reoutine, but is
applied to prevent unused caches from being loaded when it is unused by
the program. It also has other parameters for Spawn, Retail, and
Hellfire asset loading, and single-player or multiplayer.

AJenbo commented on Source/stores.cpp:
```diff
@@ -83,9 +85,10 @@ void InitStores()
 {
 	int i;
 
-	pSTextBoxCels = LoadFileInMem("Data\\TextBox2.CEL", NULL);
-	pSPentSpn2Cels = LoadFileInMem("Data\\PentSpn2.CEL", NULL);
-	pSTextSlidCels = LoadFileInMem("Data\\TextSlid.CEL", NULL);
```
stores.cpp isn't actually being build, we kept it for now in case we needed bits form it for some sort of shop analysis, but even if we do we sure aren't going to be loading these bits of UI graphics :)

AJenbo commented on Source/asset/cacheconfig.h:
```diff
@@ -0,0 +1,31 @@
+struct CacheConfig {
+	RetailMode retailMode;
+	GameMode gameMode;
+	bool ignoreUnusedAssets;
```
Hm I was hoping for less stuff to review, not more :D
